resources: Add a tutorial on adding workloads

kunpai opened the pull request:
This document details how to add a Resource, wrap it in a Workload, and test the Workload with the new gem5 Resources infrastructure introduced in v23.0.

rogerchang23424 committed:
resources: add rv32 tests to asmtest

Change-Id: Ifac87ead32638289c3f197e1fee6de290eab1a38
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/65551
Tested-by: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>

BobbyRBruce committed:
resources: Add rv32 asmtest binaries to resources.json

As of this change:
https://gem5-review.googlesource.com/c/public/gem5-resources/+/65551 we
have rv32 tests for the ASMTest. This change points to compiled binaries
hosted in our cloud bucket via the "resources.json".

Change-Id: I5aaa9be4cde82f34c3a1c1e289ddf184f1ab5c94
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/66571
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>

rogerchang23424 committed:
resources: Fix typos for rv64uc_ps_tests

Change-Id: Ic7321459b7e45966dcfa69292eb2be93e16dede7
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/66591
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: Bobby Bruce <bbruce@ucdavis.edu>
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>

rogerchang23424 committed:
resources: Add RV32/RV64 bitmanip 1.0 instruction tests

This set is folk from riscv-tests and create by myself, The test sets
include Zba, Zbb, Zbc and Zbs, is used to verify the decoder.isa

Change-Id: Id740e090090c775653265bbf0a29e83b67bb6ce3
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/66651
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: Bobby Bruce <bbruce@ucdavis.edu>
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>

rogerchang23424 committed:
resources: Update asmtest README.md

Change-Id: If6d0903930a79a963200da72a75f4ce3a1410be3
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/66812
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: Jason Lowe-Power <power.jg@gmail.com>

BobbyRBruce committed:
resources: Add RV32/RV64 inst tests to resources.json

Based on this commit:
https://gem5-review.googlesource.com/c/public/gem5-resources/+/66651
the tests were compiled and uploaded to gem5's Google Cloud Bucket. This
commit updates the resources.json file accordingly so they may be used
via the stdlib infrastructure.

Due to some confusion, the entire insttests were recompiled and
uploaded, ergo md5sum values have ben updated.

Change-Id: I95d9b67d0d4a8d6b80535c959bbba2c0a55cd5b6
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/67031
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: Bobby Bruce <bbruce@ucdavis.edu>

BobbyRBruce committed:
resources: Add 'x86-matrix-multiply-omp' resource

Change-Id: I5ec0ac757339505f4b76e7b50f0c539bfe1b7480
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/67494
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: Bobby Bruce <bbruce@ucdavis.edu>

BobbyRBruce committed:
resources: Add 'x86-matrix-multiply-omp...' resource

Change-Id: I67b64e686d53e8885d050998a46fb9dbcc7af924
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/67495
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: Bobby Bruce <bbruce@ucdavis.edu>
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: Jason Lowe-Power <power.jg@gmail.com>

BobbyRBruce committed:
resources: Add checkpoint regions for 'x86-matrix-...'

This commits adds checkpoint resources for regions 1,2,3,5,6,7,8,9,
10,11,12,13, and 14 for the 'x86-matrix-multiplt-omp' as generated in
"configs/example/gem5_library/looppoints/create-looppoint-checkpoints.py"

Change-Id: I98cc271c28e69d759aa68ccea6f59b0195f3c30f
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/67496
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: Jason Lowe-Power <power.jg@gmail.com>

BobbyRBruce committed:
resources: Add the "..omp-100-8-loopppint" resource

This commit adds the "x86-matrix-multiply-omp-100-8-looppoint" resource.

Change-Id: I9a166b579146b40a92f6d248ae079878ee1bc5de
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/67497
Tested-by: Bobby Bruce <bbruce@ucdavis.edu>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>

BobbyRBruce committed:
resources: Add Workloads for looppoint examples

Change-Id: If7a64632d032af6208ac1386ca6a3ce9fafd77ec
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/68077
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: Jason Lowe-Power <power.jg@gmail.com>

BobbyRBruce committed:
resources: Specialize the looppoint csv and Json resources

This commit specializes the Looppoint csv and JSON resources, necessary
for our example scripts to function correctly.

Note: It's still a TODO to this for the remainder of the resources.
Change-Id: I4f7a81bd3fceff98589e87ede8e23f008fbe8152
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/68137
Tested-by: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>

kunpai committed:
resources: Fixed the spelling of GAPBS "source"

The "source" field of GAPBS was "src/gabps". Fixed it
to "src/gapbs"

Change-Id: I691b633b3354b6ae82670913b734d0c9835df205
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/68158
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: Jason Lowe-Power <power.jg@gmail.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>

BobbyRBruce committed:
resources: Update matrix-multi simpoint checkpoints

These have been updated. As such their md5sum values need updated.

Change-Id: Ia45b9fb70a76a7134cf4fd32409432251ff0443f
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/68237
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: Bobby Bruce <bbruce@ucdavis.edu>

BobbyRBruce committed:
resources: Update looppoint JSON md5sum

Change-Id: Ia4dd9a0c8fffb3efdbe3f67e9b87ac2dfe48d70d
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/68278
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>

BobbyRBruce committed:
resources: Add riscv-hello-example checkpoints for v22.1.1.

Change-Id: Ie76a597d0fd59d1877e493fcd21f802cfbc7fc87
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/68279
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: Bobby Bruce <bbruce@ucdavis.edu>

BobbyRBruce committed:
resources: Fix broken URL for riscv-hello-checkpoint

Change-Id: I083894aa5ad265fe2561e2a0fe871f3371cbe795
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/68280
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: Bobby Bruce <bbruce@ucdavis.edu>
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>

BobbyRBruce committed:
resources: Add 'simpoints-se-checkpoints-v23-0-v1' resource

This is an update of the simpoints-se-checkpoints resource.

Change-Id: I5201e08becc618805b5c150df86a1a0d24fd858b
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/68317
Maintainer: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: Jason Lowe-Power <power.jg@gmail.com>
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>

BobbyRBruce committed:
resources: Rename 'riscv-hello-example-checkpoint-v22-1-1'

Rename to 'riscv-hello-example-checkpoint-v23', this resource is no
longer going to be part of a minor release.

Change-Id: I845cbd95e3e0c2b8c996c388fa52c5ea0e6bc067
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/68318
Reviewed-by: Jason Lowe-Power <power.jg@gmail.com>
Tested-by: Jason Lowe-Power <power.jg@gmail.com>
Maintainer: Jason Lowe-Power <power.jg@gmail.com>

BobbyRBruce committed:
resources: Add 'x86-gpu-square' resource

Change-Id: I12d4e5656584c7718fbd7a5d104746e3ff510171
Reviewed-on: https://gem5-review.googlesource.com/c/public/gem5-resources/+/70438
Reviewed-by: Bobby Bruce <bbruce@ucdavis.edu>
Maintainer: Bobby Bruce <bbruce@ucdavis.edu>
Tested-by: Bobby Bruce <bbruce@ucdavis.edu>

kunpai committed:
resources: Add a tutorial on adding workloads

This document details how to add a Resource, wrap it
in a Workload, and test the Workload with the new gem5
Resources infrastructure introduced in v23.0.

kunpai committed:
resources: Add a tutorial on adding workloads

This document details how to add a Resource, wrap it
in a Workload, and test the Workload with the new gem5
Resources infrastructure introduced in v23.0.

kunpai committed:
resources: Add a tutorial on adding workloads

This document details how to add a Resource, wrap it
in a Workload, and test the Workload with the new gem5
Resources infrastructure introduced in v23.0.

powerjg commented on ADDING-WORKLOADS.md:
```diff
@@ -0,0 +1,70 @@
+
+## Testing the Workload
+
+To test the Workload, we first have to add the local JSON file as a data source for gem5. This can be done by navigating to `src/python/gem5_default_config.py` and adding a JSON object to `"sources"` in the following format:
```
I really don't like this suggestion. We should not require people to modify something in source and recompile gem5 just to test a local workload.

Do we not have a way to use an environment variable to point to a json file? If not, we should add that!

kunpai replied:
According to this function we wrote: [https://github.com/gem5/gem5/blob/af72b9ba580546ac12ce05bfaac3fd53fa8699f4/src/python/gem5/resources/client.py#L53](https://github.com/gem5/gem5/blob/af72b9ba580546ac12ce05bfaac3fd53fa8699f4/src/python/gem5/resources/client.py#L53), we can either place a gem5 resources config file in the cwd or pass its path in through GEM5_CONFIG while building. So, I should indicate those two options?

powerjg replied:
Yes! Thanks :)

kunpai committed:
Merge branch 'adding-workloads-documentation' of https://github.com/kunpai/gem5-resources into adding-workloads-documentation

BobbyRBruce commented on ADDING-WORKLOADS.md:
````diff
@@ -0,0 +1,78 @@
+
+Let's assume that the Resource we want to wrap in a Workload is compiled for `RISC-V`, categorized as a `binary`, and has the name `binary-resource`. We can define this Resource in a JSON object as follows:
+
+``` json
+{
+    "category": "binary",
+    "id": "binary-resource",
+    "description": "A RISCV binary used to test a specific RISCV instruction.",
+    "architecture": "RISCV",
+    "is_zipped": false,
+    "resource_version": "1.0.0",
+    "gem5_versions": [
+        "23.0"
+    ],
+}
+```
````
How is the user supposed to set the path to the the binary resource in this case?

kunpai replied:
I can look into this, but my purpose of not mentioning it was also on the base assumption that the Resource should exist on gem5 Resources before one makes a Workload of it. Should we allow both to be done simultaneously?

With the REU cohort, they are making Workloads out of Resources that already exist or are making the Resource first before they make a Workload, which is why the tutorial follows its current flow.

BobbyRBruce replied:
I don't think this makes much sense. Perhaps I'm missing something. If you define the resource inside the local JSON file then it'll use that resource. In this case `binary-resource`. That resource is not in the database. How is gem5 supposed to know where this binary is?

BobbyRBruce commented on ADDING-WORKLOADS.md:
```diff
@@ -0,0 +1,78 @@
+
+This tutorial will walk you through the process of creating a Workload in gem5 and testing it, through the new gem5 Resources infrastructure introduced in gem5 v23.0.
+
+## Introduction
```
I would explain what a workload is. Particularly that it declaring a function call to a board.

`board.set_workload('x')` -> this. The Workload with ID 'x' will be parsed and it will be used to construction the function call it defines.

kunpai replied:
Sounds good! I'll work on it.

BobbyRBruce commented on ADDING-WORKLOADS.md:
```diff
@@ -0,0 +1,78 @@
+    "category": "workload",
+    "description": "A RISCV binary used to test a specific RISCV instruction.",
+    "architecture": "RISCV",
+    "function": "set_se_binary_workload",
```
I'd link to the `set_se_binary_workload` function here and explain the `resources` field here specifies the setting of the binary. You should also mention that the `additional_parameters` field. Why not have this binary accept some arguments?

kunpai replied:
I like this suggestion!

kunpai committed:
resources: Edit ADDING-WORKLOADS.md

Some clarifications on:

- set_workload function
- set_se_binary_workload functions

powerjg reviewed (changes requested):
Before merging, can you clean up the git history (i.e., rebase and/or squash) into a human-understandable history?

kunpai committed:
Merge branch 'adding-workloads-documentation' of https://github.com/kunpai/gem5-resources into adding-workloads-documentation

BobbyRBruce commented:
Did you mean to close this?

kunpai commented:
Yes, I need to edit the tutorial. I will reopen it once Workloads are deprecated and my change on supporting paths is merged
